bokehjs/src/lib/models/widgets/tables/data_cube.ts

```diff
@@ -0,0 +1,360 @@
+  export type Attrs = p.AttrsOf<Props>
+
+  export type Props = Model.Props & {
+    getter:      p.Property<string>
+    aggregators: p.Property<RowAggregator[]>
+    collapsed:   p.Property<boolean>
+  }
+}
+
+export interface GroupingInfo extends GroupingInfo.Attrs {}
+
+export class GroupingInfo extends Model {
+  properties: GroupingInfo.Props
+
+  constructor(attrs?: Partial<GroupingInfo.Attrs>) {
+    super(attrs)
+  }
+
+  static initClass(): void {
+    this.prototype.type = 'GroupingInfo'
+
+    this.define<GroupingInfo.Props>({
+      getter:      [ p.String,   ''    ],
+      aggregators: [ p.Array,    []    ],
+      collapsed:   [ p.Boolean,  false ],
+    })
+  }
+
+  get comparer(): (a: { value: any }, b: { value: any }) => number {
+    return (a: { value: any }, b: { value: any }): number => {
+      return a.value === b.value ? 0 : a.value > b.value ? 1 : -1
+    }
+  }
+}
+GroupingInfo.initClass()
+
+export class DataCubeProvider extends DataProvider {
+
+  readonly columns: Column[]
+  groupingInfos: GroupingInfo[]
+  readonly groupingDelimiter: string
+  toggledGroupsByLevel: {[key: string]: boolean}[]
+  private rows: (Group | number)[]
+  target: ColumnDataSource
+
+  constructor(source: ColumnDataSource, view: CDSView, columns: Column[], target: ColumnDataSource) {
+    super(source, view)
+    this.columns = columns
+    this.groupingInfos = []
+    this.groupingDelimiter = ':|:'
+    this.target = target
+  }
+
+  setGrouping(groupingInfos: GroupingInfo[]): void {
+    this.groupingInfos = groupingInfos
+    this.toggledGroupsByLevel = groupingInfos.map(() => ({}))
+
+    this.refresh()
+  }
+
+  private extractGroups(rows: number[], parentGroup?: Group): Group[] {
+    const groups: Group[] = []
+    const groupsByValue: Map<any, Group> = new Map()
+    const level = parentGroup ? parentGroup.level + 1 : 0
+    const { comparer, getter } = this.groupingInfos[level]
+
+    rows.forEach((row: number) => {
+      const value = this.source.data[getter][row]
+      let group = groupsByValue.get(value)
+
+      if (!group) {
+        const groupingKey = parentGroup ? `${parentGroup.groupingKey}${this.groupingDelimiter}${value}` : `${value}`
+        group = Object.assign(new Group(), {
+          value,
+          level,
+          groupingKey,
+        })
+        groups.push(group)
+        groupsByValue.set(value, group)
+      }
+      group.rows.push(row)
+    })
+
+    if (level < this.groupingInfos.length - 1) {
+      groups.forEach((group) => {
+        group.groups = this.extractGroups(group.rows, group)
+      })
+    }
+
+    groups.sort(comparer)
+    return groups
+  }
+
+  private calculateTotals(group: Group, aggregators: RowAggregator[]): GroupTotals {
+    const totals: GroupTotals = { avg: {}, max: {}, min: {}, sum: {} }
+    const { source: { data } } = this
+    const keys = Object.keys(data)
+    const items = group.rows.map((i: number) => keys.reduce((o, c) => ({ ...o, [c]: data[c][i] }), {}))
```

Comment:
In general we prefer inner arrow functions not to use argument type declarations, but instead to use type inference. If the compiler complains about missing/implicit types, then it's often a good indicator that either `any` sneaked in somewhere or the type of `group.rows` is messed up in some other way.